cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -253,22 +253,21 @@ func AddNewControlPlaneTaint(client clientset.Interface) error {
 	}
 
 	for _, n := range nodes.Items {
-		// Check if the node has the old / new taints
+		// Check if the node has the old taint
 		hasOldTaint := false
-		hasNewTaint := false
+		taints := []v1.Taint{}
 		for _, t := range n.Spec.Taints {
-			switch t.String() {
-			case kubeadmconstants.OldControlPlaneTaint.String():
+			if t.String() == kubeadmconstants.OldControlPlaneTaint.String() {
 				hasOldTaint = true
-			case kubeadmconstants.ControlPlaneTaint.String():
-				hasNewTaint = true
+				continue
 			}
+			// Collect all other taints
+			taints = append(taints, t)
 		}
-		// If the old taint is present and the new taint is missing, patch the node with the new taint.
-		// When the old taint is missing, assume the user has manually untainted the node and take no action.
-		if !hasNewTaint && hasOldTaint {
+		// If the old taint is present remove it
```

Comment:
If someone manually removes the new taint before they upgrade to v1.25, are we assured that the new taint is put back before we remove the old taint?

(I'd expected to see at least a warning that the old taint is being removed but the new taint isn't present).

Reply:
> If someone manually removes the new taint before they upgrade to v1.25, are we assured that the new taint is put back before we remove the old taint?

I personally don't expect this use case. If someone manually removed the new taint we can assume they wanted it removed and we should not put it back in. There was a related bug report that kubeadm should not reintroduce removed taints - e.g. in single control plane node clusters.

> (I'd expected to see at least a warning that the old taint is being removed but the new taint isn't present).

The output of "upgrade" will print a message that the old taint is being removed. Combined with the release note it should give sufficient awareness to users.

Reply:
Cool, makes sense.